docs/adr/0003-delivery-channels-templating.md

```diff
@@ -0,0 +1,92 @@
+
+Creation of templates can be totally flexible and dependent on the organisation which will integrate with `Marain.UserNotifications` thus every tenant will have their own set of templates for SMS and Email etc. Default templates will be provided at the start for the communication types (SMS, email, web push notifications) at start but can be modified and updated which will leave space for the organisation to configure as per their need.
+
+### Storage of templates
+
+Templates need to be stored in such a way where it can be accessed and modified easily on the fly. Thus to have this flexibilty templates will be stored in azure table storage. The template for individual properties (`ToPhoneNumber`, `FromPhoneNumber`, `ToEmail`) would have the following column headers:
+
+- Notification Type
+- Communication Type
+- Property Name (This will be the property name for the the communication object)
+- Template
+
+Eg. The below example shows how the table storage will host the relevant templates mentioned above having a template for every single property name per communication type and notification type.
+
+```
+Notification Type       Communication Type         Property Name        Template
+
+New Lead                SMS                        ToPhoneNumber        {{toPhoneNumber}}
+
+New Lead                SMS                        Body                 Hi {{endUserName}},
+                                                                        You have been assigned a new lead from {{fromUserName}}.
+
+New Lead                Email                      ToEmail              {{toEmail}}
+
+New Lead                Email                      Body                 Hi {{endUserName}},
+                                                                        You have been assigned a new lead from {{fromUserName}} having name {{leadName}}.
+
+```
+
+### Applying a certain template to a property for a communication object
+
+The Management API will be responsible to convert the create notification request into a particular communication object (SMSObject) based on the configured user preferences for the end user. Every communication object has certain required properties and without which a notification cannot be sent as explained in the section above. If we arrive in such a scenario where the essential properties are missing then those can be logged in App Insights detailing which property name was missing for a notification type and communication type and who was the end user who triggered this notification.
```

Comment:
Typo:

> Every communication object has certain required properties **and without** which 

I think you don't want that "and" in there, unless I've misunderstood the sentence (in which case, please clarify it).


>those can be logged in App Insights 

Bear in mind that Application Insights drops information when it gets busy, so this is not a reliable way of reporting failure. More generally, it's also fairly easy to ignore. If a problem of the kind described exists, we've got a big problem because a whole class of communications is now going to fail repeatedly. Ideally this would be indicating to the caller in a way that is harder to ignore.

Have you considered the possibility of returning some sort of failure result if the service is incorrectly configured, so that the caller knows that their notification is definitely not going to be delivered?

Reply:
Agreed. We would be storing it to the existing table with the status and reason if failed. Updated the ADR accordingly.